File: src/MagnusAddon/jaslib/note/vocabulary/related_vocab/SeeAlso.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from jaslib import app
from jaslib.app import col

if TYPE_CHECKING:
    from jaslib.note.notefields.json_object_field import MutableSerializedObjectField
    from jaslib.note.vocabulary.related_vocab.related_vocab_data import RelatedVocabData
    from jaslib.note.vocabulary.vocabnote import VocabNote
    from jaslib.sysutils.weak_ref import WeakRef
    from typed_linq_collections.collections.q_set import QSet


class SeeAlso(Slots):
    def __init__(self, vocab: WeakRef[VocabNote], data: MutableSerializedObjectField[RelatedVocabData]) -> None:
        self._vocab: WeakRef[VocabNote] = vocab
        self._data: MutableSerializedObjectField[RelatedVocabData] = data

    def strings(self) -> QSet[str]: return self._data.get().see_also
    def notes(self) -> list[VocabNote]:
        return col().vocab.with_any_form_in_prefer_disambiguation_name_or_exact_match(list(self.strings()))

    def add(self, to_add: str) -> None:
        self.strings().add(to_add)

        for added_note in app.col().vocab.with_question(to_add):
            if self._vocab().get_question() not in added_note.related_notes.see_also.strings():
                added_note.related_notes.see_also.add(self._vocab().get_question())

        self._data.save()

    def remove(self, to_remove: str) -> None:
        self.strings().remove(to_remove)

        for removed_note in app.col().vocab.with_question(to_remove):
            if self._vocab().get_question() in removed_note.related_notes.see_also.strings():
                removed_note.related_notes.see_also.remove(self._vocab().get_question())

        self._data.save()

    @override
    def __repr__(self) -> str: return self._data.__repr__()

File: src/src_dotnet/JAStudio.Core/Note/Vocabulary/RelatedVocab/SeeAlso.cs
using System.Collections.Generic;
using System.Linq;

namespace JAStudio.Core.Note.Vocabulary.RelatedVocab;

public class SeeAlso
{
   readonly VocabNote _vocab;
   readonly RelatedVocabData _data;
   readonly NoteGuard _guard;

   public SeeAlso(VocabNote vocab, RelatedVocabData data, NoteGuard guard)
   {
      _vocab = vocab;
      _data = data;
      _guard = guard;
   }

   public HashSet<string> Strings() => _data.SeeAlso;

   public List<VocabNote> Notes() => _vocab.Services.Collection.Vocab.WithAnyFormInPreferDisambiguationNameOrExactMatch(Strings().ToList());

   public void Add(string toAdd) => _guard.Update(() =>
   {
      Strings().Add(toAdd);

      foreach(var addedNote in _vocab.Services.Collection.Vocab.WithQuestion(toAdd))
      {
         if(!addedNote.RelatedNotes.SeeAlso.Strings().Contains(_vocab.GetQuestion()))
         {
            addedNote.RelatedNotes.SeeAlso.Add(_vocab.GetQuestion());
         }
      }
   });

   public void Remove(string toRemove) => _guard.Update(() =>
   {
      Strings().Remove(toRemove);

      foreach(var removedNote in _vocab.Services.Collection.Vocab.WithQuestion(toRemove))
      {
         if(removedNote.RelatedNotes.SeeAlso.Strings().Contains(_vocab.GetQuestion()))
         {
            removedNote.RelatedNotes.SeeAlso.Remove(_vocab.GetQuestion());
         }
      }
   });

   public override string ToString() => string.Join(", ", Strings());
}
